src/acl.c

```diff
@@ -934,6 +958,93 @@ static void ACLAddAllowedFirstArg(aclSelector *selector, unsigned long id, const
     selector->allowed_firstargs[id][items - 2] = sdsnew(sub);
     selector->allowed_firstargs[id][items - 1] = NULL;
 }
+/* Initialize intset for database permissions. */
+static intset *ACLInitDatabaseIntset(aclSelector *selector, int add_dbs) {
+    intset *new_dbs;
+
+    /* For removal operations with ALLDBS flag, convert to explicit list */
+    if (!add_dbs && (selector->flags & SELECTOR_FLAG_ALLDBS)) {
+        new_dbs = intsetNew();
+        if (!new_dbs) return NULL;
+
+        for (int i = 0; i < server.dbnum; i++) {
+            uint8_t success;
+            new_dbs = intsetAdd(new_dbs, i, &success);
+            serverAssert(success);
+        }
+        return new_dbs;
+    }
+
+    return selector->dbs ? intsetDup(selector->dbs) : intsetNew();
+}
+
+/* Clean up and return error from database permission operations. */
+static int ACLDatabasePermissionError(intset *new_dbs, char *dblist, int error_code) {
+    if (new_dbs) intsetFree(new_dbs);
+    if (dblist) zfree(dblist);
+    errno = error_code;
+    return C_ERR;
+}
+
+/* Helper function to add or remove database permissions to/from a selector.
+ * When add_dbs is 1, this function adds the specified databases to the existing set.
+ * When add_dbs is 0, this function removes the specified databases from the existing set. */
+static int ACLSetSelectorDatabasePermissions(aclSelector *selector, const char *dbs_str, int add_dbs) {
+    /* Initialize the database intset based on operation type */
+    intset *new_dbs = ACLInitDatabaseIntset(selector, add_dbs);
+    if (!new_dbs) {
+        return ACLDatabasePermissionError(NULL, NULL, ENOMEM);
+    }
```

Comment:
I think the only scenario in which you return `NULL` isn't required. So, we can get rid of this as well.


```suggestion
```